packages/devtools_app/lib/src/flutter/banner_messages.dart

```diff
@@ -0,0 +1,304 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/material.dart';
+
+import '../auto_dispose.dart';
+import 'auto_dispose_mixin.dart';
+import 'common_widgets.dart';
+import 'controllers.dart';
+import 'screen.dart';
+import 'theme.dart';
+import 'utils.dart';
+
+const _runInProfileModeDocsUrl =
+    'https://flutter.dev/docs/testing/ui-performance#run-in-profile-mode';
+
+const _profileGranularityDocsUrl =
+    'https://flutter.dev/docs/development/tools/devtools/performance#profile-granularity';
+
+class BannerMessagesController implements DisposableController {
+  final _dismissedMessageIds = <String>{};
+
+  final _refreshMessagesController = StreamController<Null>.broadcast();
+
+  Stream<Null> get onRefreshMessages => _refreshMessagesController.stream;
+
+  @override
+  void dispose() {
+    _refreshMessagesController.close();
+  }
+
+  void refreshMessages() {
+    _refreshMessagesController.add(null);
+  }
+
+  void dismissMessage(String messageId) {
+    assert(!isMessageDismissed(messageId));
+    _dismissedMessageIds.add(messageId);
+    refreshMessages();
+  }
+
+  bool isMessageDismissed(String messageId) {
+    return _dismissedMessageIds.contains(messageId);
+  }
+}
+
+class BannerMessageContainer extends StatefulWidget {
+  const BannerMessageContainer({Key key, @required this.screen})
+      : super(key: key);
+
+  final Screen screen;
+
+  @override
+  _BannerMessageContainerState createState() => _BannerMessageContainerState();
+}
+
+class _BannerMessageContainerState extends State<BannerMessageContainer>
+    with AutoDisposeMixin {
+  BannerMessagesController controller;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    final newController = Controllers.of(context)?.bannerMessages;
+    if (newController == controller) return;
+    controller = newController;
+
+    autoDispose(controller.onRefreshMessages.listen((_) => setState(() {})));
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final messagesForScreen = widget.screen.messages(context);
+    if (messagesForScreen.isNotEmpty) {
+      final messagesToShow = <Widget>[];
+      for (Widget message in messagesForScreen) {
+        assert(message is UniqueMessage);
```

Comment:
if we only handle UniqueMessage objects, change the type to
`List<UniqueMessage>`

Reply:
you'll need to change UniqueMessage to a MessageWidget base class that has an id.
UniqueMessage is a bit of a confusing to me as it doesn't have any members related to being a message so that is probably a good thing.